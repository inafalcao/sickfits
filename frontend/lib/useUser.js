import { useMutation, useQuery } from '@apollo/client';
import gql from 'graphql-tag';
import { useRouter } from 'next/dist/client/router';

export const CURRENT_USER_QUERY = gql`
  query {
    authenticatedItem {
      ... on User {
        id
        email
        name
      }
    }
  }
`;

const SIGN_OUT_MUTATION = gql`
  mutation {
    endSession
  }
`;

export default function useUser() {
  const { data } = useQuery(CURRENT_USER_QUERY);
  const [endSession] = useMutation(SIGN_OUT_MUTATION, {
    refetchQueries: [{ query: CURRENT_USER_QUERY }],
  });
  const router = useRouter();

  async function signOut() {
    await endSession();
    router.push('/signin');
  }

  return { user: data?.authenticatedItem, signOut };
}
