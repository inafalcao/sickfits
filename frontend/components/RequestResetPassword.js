import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import { useRouter } from 'next/dist/client/router';
import Form from './styles/Form';
import useForm from '../lib/useForm';
import { CURRENT_USER_QUERY } from '../lib/useUser';
import DisplayError from './ErrorMessage';

export const REQUEST_RESET_MUTATION = gql`
  mutation REQUEST_RESET_MUTATION($email: String!) {
    sendUserPasswordResetLink(email: $email) {
      code
      message
    }
  }
`;

export default function RequestResetPassword() {
  const { inputs, handleChange } = useForm({ email: '' });

  const [sendUserPasswordResetLink, { loading, data, error }] = useMutation(
    REQUEST_RESET_MUTATION,
    {
      variables: inputs,
    }
  );

  async function handleSubmit(e) {
    e.preventDefault();

    const res = await sendUserPasswordResetLink().catch(console.error);
    console.log(res);
  }

  return (
    <Form method="POST" onSubmit={handleSubmit}>
      <h2>Request Reset Password</h2>
      <DisplayError error={error} />
      {data?.sendUserPasswordResetLink === null && (
        <p>Success! Check your e-mail for a link.</p>
      )}
      <fieldset>
        <label htmlFor="email">
          Email
          <input
            type="email"
            name="email"
            placeholder="Email"
            autoComplete="email"
            value={inputs.email}
            onChange={handleChange}
          />
        </label>

        <button type="submit">Request reset</button>
      </fieldset>
    </Form>
  );
}
