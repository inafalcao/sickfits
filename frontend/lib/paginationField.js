import { TOTAL_PRODUCTS_QUERY } from '../components/Pagination';

export default function paginationField() {
  return {
    keyArgs: false, // Tells Apollo we will take care of everything
    read(existing = [], { args, cache }) {
      // If we already have the field, return it.
      // If we return false, it will do a network request
      const { skip, first } = args;

      const data = cache.readQuery({ query: TOTAL_PRODUCTS_QUERY });

      const count = data?._allProductsMeta?.count;
      const page = skip / first + 1;
      const pages = Math.ceil(count / first);

      const items = existing.slice(skip, skip + first).filter((x) => x);

      if (
        (page === pages && items.length && items.length < first) ||
        items.length === first
      ) {
        return items;
      }

      if (items.length !== first) {
        return false;
      }

      return false;
    },
    merge(existing, incoming, { args }) {
      // This runs when the Apollo client comes back from the network with our fields
      // Here we say how we're going to put them on the cache
      const { skip } = args;

      const merged = existing ? [...existing] : [];

      for (let i = skip; i < skip + incoming.length; i++) {
        merged[i] = incoming[i - skip];
      }

      return merged;
    },
  };
}
