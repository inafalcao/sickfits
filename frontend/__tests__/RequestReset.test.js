import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import userEvent from '@testing-library/user-event';
import { fakeUser } from '../lib/testUtils';
import RequestResetPassword, {
  REQUEST_RESET_MUTATION,
} from '../components/RequestResetPassword';

const me = fakeUser();
const mocks = [
  {
    request: {
      query: REQUEST_RESET_MUTATION,
      variables: { email: me.email },
    },
    result: {
      data: { sendUserPasswordResetLink: null },
    },
  },
];

describe('<RequestReset/>', () => {
  it('renders and matches snapshot', () => {
    const { container } = render(
      <MockedProvider>
        <RequestResetPassword />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('calls the mutation when submited', async () => {
    render(
      <MockedProvider mocks={mocks}>
        <RequestResetPassword />
      </MockedProvider>
    );
    await userEvent.type(screen.getByPlaceholderText(/Email/i), me.email);

    await userEvent.click(screen.getByText('Request reset'));
    await screen.findByText('Success! Check your e-mail for a link.');
  });
});
