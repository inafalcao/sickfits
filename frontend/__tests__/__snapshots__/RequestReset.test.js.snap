// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestReset/> renders and matches snapshot 1`] = `
<div>
  <form
    class="Form-sc-1xszr8q-0 hVEEdl"
    method="POST"
  >
    <h2>
      Request Reset Password
    </h2>
    <fieldset>
      <label
        for="email"
      >
        Email
        <input
          autocomplete="email"
          name="email"
          placeholder="Email"
          type="email"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Request reset
      </button>
    </fieldset>
  </form>
</div>
`;
